README.md
![bewegung_logo](https://github.com/pleiszenburg/bewegung/blob/master/docs/_static/logo01.png?raw=true "bewegung logo")

# BEWEGUNG

*/bəˈveːɡʊŋ/ ([German, noun, feminine: motion/movement/animation](https://dict.leo.org/englisch-deutsch/bewegung))*

[![docs_master](https://readthedocs.org/projects/bewegung/badge/?version=latest&style=flat-square "Documentation Status: master / release")](https://bewegung.readthedocs.io/en/latest/)
[![license](https://img.shields.io/pypi/l/bewegung.svg?style=flat-square "LGPL 2.1")](https://github.com/pleiszenburg/bewegung/blob/master/LICENSE)
[![status](https://img.shields.io/pypi/status/bewegung.svg?style=flat-square "Project Development Status")](https://github.com/pleiszenburg/bewegung/issues)
[![pypi_version](https://img.shields.io/pypi/v/bewegung.svg?style=flat-square "pypi version")](https://pypi.python.org/pypi/bewegung)
[![conda_version](https://img.shields.io/conda/vn/conda-forge/bewegung.svg?style=flat-square "conda version")](https://anaconda.org/conda-forge/bewegung)
[![pypi_versions](https://img.shields.io/pypi/pyversions/bewegung.svg?style=flat-square "Available on PyPi - the Python Package Index")](https://pypi.python.org/pypi/bewegung)
[![chat](https://img.shields.io/matrix/bewegung:matrix.org.svg?style=flat-square "Matrix Chat Room")](https://matrix.to/#/#bewegung:matrix.org)
[![mailing_list](https://img.shields.io/badge/mailing%20list-groups.io-8cbcd1.svg?style=flat-square "Mailing List")](https://groups.io/g/bewegung-dev)

## Synopsis

`bewegung` is a versatile video renderer, primarily targeting scientific visualizations of large quantities of data. Its core concepts are *sequences* and *layers*. Sequences describe a certain time span within a video and can overlap. Each sequence can hold multiple layers. Layers can be generated with [`cairo`](https://cairographics.org/), [`Pillow`](https://pillow.readthedocs.io), [`datashader`](https://datashader.org/), [`matplotlib`](https://matplotlib.org/) and `bewegung`'s internal drawing system [`DrawingBoard`](https://bewegung.readthedocs.io/en/latest/canvas.html). Final compositing of every video frame and video effects are implemented via `Pillow`. Video encoding is handled by `ffmpeg`. `bewegung` also includes a simple [vector algebra system](https://bewegung.readthedocs.io/en/latest/vectors.html) and a ["camera" for 3D to 2D projections](https://bewegung.readthedocs.io/en/latest/camera.html). `bewegung` is developed with ease of use, compute time and memory efficiency in mind.

## Installation

`bewegung` can be installed both via ``conda`` and via ``pip``.

### Via `conda`

An almost complete installation can be triggered by running:

```bash
conda install -c conda-forge bewegung
```

Please note that [mplcairo](https://github.com/matplotlib/mplcairo), a dependency of `bewegung` and alternative backend for `matplotlib`, is currently not available via `conda` and must be installed manually. `bewegung` [does also work without `mplcairo` present](https://bewegung.readthedocs.io/en/latest/canvas.html#acceleratingmatplotlib) and falls back to the `cairo` backend of `matplotlib`.

### Via `pip`

A bare **minimum** of `bewegung` can be installed with Python's package manager `pip`:

```bash
pip install -vU bewegung
```

A **complete** installation of all optional Python components and development tools can be triggered by running:

```bash
pip install -vU bewegung[all]
```

Certain non-Python **prerequisites** must installed separately and before invoking the above command. [For detailed instructions, see documentation](https://bewegung.readthedocs.io/en/latest/installation.html). Most notably, `ffmpeg` should be installed for producing actual video files instead of video frames as individual files. See [download section](https://ffmpeg.org/download.html) of the `ffmpeg` project website for further instructions.

## Example

See [`demo.py`](https://github.com/pleiszenburg/bewegung/blob/master/demo/demo.py).

You can directly test it by running:

```bash
curl https://raw.githubusercontent.com/pleiszenburg/bewegung/master/demo/demo.py | python3
```

This resulting `video.mp4` file should look like this:

[![bewegung standard demo](https://img.youtube.com/vi/4NFXQ73weMA/sddefault.jpg)](https://www.youtube.com/watch?v=4NFXQ73weMA)

## Usage

See [documentation](https://bewegung.readthedocs.io).

`bewegung`'s development status is "well-tested alpha". Its API should not be considered stable until the project is labeled "beta" or better.

`bewegung` can be drastically accelerated by deactivating debugging features. See [relevant section in the documentation](https://bewegung.readthedocs.io/en/latest/performance.html#typecheckingperformance).


CHANGES.md
# Changes

## 0.0.8 (2022-XX-XX)

- (TBD)

## 0.0.7 (2022-03-27)

- DOCS: Updated "Getting Started" chapter to showcase new meta data dictionary in vector and matrix classes.
- DOCS: Bumped Python at ReadTheDocs from 3.7 to 3.8.
- FIX: Docs would not build properly with Python 3.8 and older.

## 0.0.6 (2021-11-07)

Highlights: Major overhaul of linear algebra functionality, better package structure and a test suite.

- FEATURE: All vector and vector array classes expose `ndim`, number of dimensions.
- FEATURE: Common base class, `Vector`, for all vector classes.
- FEATURE: Common base class, `VectorArray`, for all vector array classes.
- FEATURE: Vector arrays are iterators.
- FEATURE: Added missing right-hand-side operators to `Vector` and `VectorArray` classes.
- FEATURE: Tuple export of `VectorArray` types can optionally provide direct access to underlying ``ndarray``s, i.e. new ``copy`` parameter can be set to ``False``.
- FEATURE: 3D vectors and vector arrays can export geographic coordinates.
- FEATURE: The `Color` class, using RGBA internally, can now import HSV values.
- FEATURE: Added equality check, "is close" check, tuple export and copy to `Matrix`.
- FEATURE: Added new `MatrixArray` class.
- FEATURE: New dedicated sub-module for core animation engine named `bewegung.animation`.
- FEATURE: New dedicated sub-module for `DrawingBoard` named `bewegung.drawingboard`, now allowing direct import.
- FEATURE: New dedicated sub-module for linear algebra named `bewegung.lingalg`.
- FEATURE: All linear algebra classes have consistent dtype and error handling.
- FEATURE: Cleanup of internal type hierarchy.
- FEATURE: Added test suite with some initial tests, based on `pytest`, `hypothesis` and `coverage`.
- API CHANGE: Vector array method `update_from_vector` renamed to `update_from_vectorarray`.
- API CHANGE: `Vector2Ddist` and `VectorArray2Ddist` removed in favor of meta data dictionaries within all vector, vector array, matrix and matrix array classes.
- FIX: Development dependency switched from unmaintained `python-language-server` to maintained fork `python-lsp-server`.
- FIX: Imports in `contrib` were broken.
- FIX: `test` target in `makefile` was broken.
- FIX: `typeguard` was not really an optional dependency.

## 0.0.5 (2021-07-30)

- FEATURE: Python 3.9 support.
- FEATURE: Added `draw_bezier` method to `DrawgingBoard`.
- FEATURE: `Matrix` can rotate vector arrays.
- FEATURE: Added operations (add, subtract) between vectors and vector arrays.
- FEATURE: Vectors and vector arrays expose angles.
- FEATURE: Matrix chat room for support.
- FEATURE: Groups.io mailing list for support.
- FIX: Some text anchors would fail with activated type checking.
- DOCS: Added project logo.
- DOCS: Small corrections in various places.

## 0.0.4 (2020-12-14)

- FIX: Center offset in `DrawgingBoard` cares about subpixels.
- FIX: `bewegung` would fail to work at all without `numpy` present.
- DOCS: Completed vector chapters on algebra and camera as well as cross-references to classes and methods.

## 0.0.3 (2020-12-06)

- FEATURE: `DrawingBoard.make_svg` can generate SVG object handles from raw binary data.
- FEATURE: Demos are self-contained.
- FIX: `DrawingBoard.draw_svg` can draw raw SVG data without crashing.
- DOCS: Prerequisites when installing via `pip`

## 0.0.2 (2020-12-05)

- FEATURE: `mplcairo` becomes an optional dependency. The `matplotlib` backend can fall back to its own `cairo` backend while also showing a warning.
- DOCS: Package installation via `conda`

## 0.0.1 (2020-12-02)

- Initial release.


demo/demo.py
# -*- coding: utf-8 -*-

"""

BEWEGUNG
a versatile video renderer
https://github.com/pleiszenburg/bewegung

    demo/demo.py: Package demo

    Copyright (C) 2020-2021 Sebastian M. Ernst <ernst@pleiszenburg.de>

<LICENSE_BLOCK>
The contents of this file are subject to the GNU Lesser General Public License
Version 2.1 ("LGPL" or "License"). You may not use this file except in
compliance with the License. You may obtain a copy of the License at
https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
https://github.com/pleiszenburg/bewegung/blob/master/LICENSE

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
specific language governing rights and limitations under the License.
</LICENSE_BLOCK>

"""

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# IMPORT
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

import math

try:
    import psutil
    CPU_CORES = psutil.cpu_count(logical = False)
except ModuleNotFoundError: # fallback
    import multiprocessing
    CPU_CORES = multiprocessing.cpu_count() // 2 # assume HT
    CPU_CORES = 1 if CPU_CORES < 1 else CPU_CORES

from bewegung import (
    Camera, Color, Video,
    Vector2D, Vector3D, VectorArray3D,
    FadeInEffect, FadeOutEffect,
    )

from bewegung.drawingboard import DrawingBoard

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# CONST
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

SVG_LOGO = """<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
   viewBox="0 0 170 170"
   width="170"
   height="170"
>
  <path
     style="fill:#b4b4b4;fill-opacity:1;"
     d="m 275.57824,185.64694 -154.01,-67.09663 135.11238,-99.828253 z"
     transform="rotate(-44.941938,124.56725,280.97355)" />
</svg>
""".encode('utf-8')

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# ROUTINES
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

def main():

    v = Video(
        width = 1920,
        height = 1080,
        seconds = 10.0,
        ctx = {
            'bg_color': Color(26, 26, 26),
            'bg_color_transparent': Color(26, 26, 26, 0),
            'camera_dist': 30,
            'camera': Camera(
                position = Vector3D(30.0, 0.0, 0.0),
                direction = Vector3D(-1.0, 0.0, 0.0),
                planeOffset = Vector2D(1920 / 2, 1080 / 2),
                planeFactor = 1000.0,
            ),
        },
    )

    @v.sequence()
    class Background:

        @v.layer(
            zindex = v.zindex.on_bottom(),
            canvas = v.canvas(background_color = v.ctx['bg_color']),
        )
        def empty(self, canvas):
            return canvas

    @v.sequence(
        start = v.time_from_seconds(1.0),
        stop = v.time_from_seconds(-1.0),
    )
    class Sphere:

        def __init__(self):
            self._lines3d = [
                VectorArray3D.from_iterable([
                    Vector3D.from_geographic(10.0, float(lon), float(lat))
                    for lat in range(-90, 90 + 10, 10)
                ])
                for lon in range(-180, 180, 10)
            ]
            self._lines3d.extend([
                VectorArray3D.from_iterable([
                    Vector3D.from_geographic(10.0, float(lon), float(lat))
                    for lon in range(-180, 180 + 10, 10)
                ])
                for lat in range(-90 + 10, 90, 10)
            ])
            self._lines2d = None
            self._factor = None

        @v.prepare(
            preporder = v.preporder.on_bottom(),
        )
        def move_camera(self, time):
            angle = 2 * math.pi * time.seconds / 120.0
            position2d = Vector2D.from_polar(radius = v.ctx['camera_dist'], angle = angle)
            direction2d = Vector2D.from_polar(radius = 1.0, angle = math.pi + angle)
            v.ctx['camera'].position = Vector3D(x = position2d.x, y = position2d.y, z = 0.0)
            v.ctx['camera'].direction = Vector3D(x = direction2d.x, y = direction2d.y, z = 0.0)
            v.ctx['camera'].roll = angle

        @v.prepare(
            preporder = v.preporder.on_top(),
        )
        def project(self):
            self._lines2d = [
                v.ctx['camera'].get_points(line3d).as_list()
                for line3d in self._lines3d
            ]
            self._lines2d = [
                (a, b)
                for line2d in self._lines2d
                for a, b in zip(line2d[:-1], line2d[1:])
            ]
            self._lines2d.sort(key = lambda item: item[0].meta['dist'], reverse = True)
            minimum = self._lines2d[0][0].meta['dist']
            maximum = self._lines2d[-1][0].meta['dist']
            self._factor = lambda x: (x - minimum) / (maximum - minimum)

        @FadeInEffect(v.time_from_seconds(4.0))
        @FadeOutEffect(v.time_from_seconds(2.0))
        @v.layer(
            zindex = v.zindex.on_top(),
            canvas = v.canvas(background_color = v.ctx['bg_color_transparent']),
        )
        def wiremesh(self, canvas):
            for line in self._lines2d:
                factor = self._factor(line[0].meta['dist'])
                gray = round(64 + 191 * factor)
                canvas.draw_polygon(
                    *line,
                    line_color = Color(gray, gray, gray),
                    line_width = 1.6 + 1.6 * factor,
                )
            return canvas

    @v.sequence(
        stop = v.time_from_seconds(3.0),
    )
    class Intro:

        def __init__(self):

            self._font = DrawingBoard.make_font('Arial', 40.0)

        @FadeInEffect(v.time_from_seconds(1.6))
        @FadeOutEffect(v.time_from_seconds(0.8))
        @v.layer(
            zindex = v.zindex.on_top(),
            canvas = v.canvas(background_color = v.ctx['bg_color_transparent'], height = 200),
            offset = Vector2D(20, 20),
        )
        def text(self, canvas):

            canvas.draw_text(
                text = 'bewegung - a versatile video renderer\nDEMO',
                point = Vector2D(0.0, 0.0),
                font = self._font,
                font_color = Color(180, 180, 180),
                anchor = 'tl',
            )

            return canvas

    @v.sequence(
        start = v.time_from_seconds(-5.0),
        stop = v.time_from_seconds(-1.0),
    )
    class Credits:

        def __init__(self):

            self._font = DrawingBoard.make_font('Arial', 40.0)

        @FadeInEffect(v.time_from_seconds(2.0))
        @FadeOutEffect(v.time_from_seconds(2.0))
        @v.layer(
            zindex = v.zindex.on_top(),
            canvas = v.canvas(background_color = v.ctx['bg_color_transparent'], height = 200),
            offset = Vector2D(-20, v.height - 20 - 200),
        )
        def text(self, canvas):

            canvas.draw_text(
                text = 'github.com/pleiszenburg/bewegung',
                point = Vector2D(v.width, 200),
                font = self._font,
                font_color = Color(180, 180, 180),
                alignment = 'r',
                anchor = 'br',
            )

            return canvas

    @v.sequence(
        start = v.time_from_seconds(-1.2),
    )
    class CreditsLogo:

        def __init__(self):

            self._svg = DrawingBoard.make_svg(raw = SVG_LOGO)
            self._font = DrawingBoard.make_font('Arial', 25.0)

        @FadeInEffect(v.time_from_seconds(0.8))
        @FadeOutEffect(v.time_from_seconds(0.3))
        @v.layer(
            zindex = v.zindex.on_top(),
            canvas = v.canvas(background_color = v.ctx['bg_color_transparent']),
        )
        def text(self, canvas):

            canvas.draw_svg(
                svg = self._svg,
                point = Vector2D(v.width / 2, v.height / 2 - 50),
                scale = 2.0,
                anchor = 'cc',
            )
            canvas.draw_text(
                text = 'pleiszenburg.de - Independent Scientific Services\n2020',
                point = Vector2D(v.width / 2, v.height / 2 + 135),
                font = self._font,
                font_color = Color(180, 180, 180),
                alignment = 'c',
                anchor = 'tc',
            )

            return canvas

    v.render(
        processes = CPU_CORES,
        video_fn = 'video.mp4',
        )

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# ENTRY
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

if __name__ == '__main__':

    main()
